### PR TITLE
Add token exposed processing to tokens server

### PR DESCRIPTION
**Pull request by gjcthinkst**
## Proposed changes

As part of a bigger change to add exposed key monitoring to AWS keys, this PR adds the functionality to the tokens server to accept POST requests (that will come from a lambda) to add a token exposed event to a token. This information is then displayed on the frontend.

This PR also adds support for this new type of alert (an exposed key alert) in the generic webhook channel, but leaves the implementation of the exposed key alert for other channels (email and the other webhooks) for a future PR to keep this PR small (we can do this since the lambda that will create these POST requests doesn't exist yet so these events won't be created under normal circumstances yet).

## Types of changes

What types of changes does your code introduce to this repository?

- [x] New feature (non-breaking change which adds functionality)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] Lint and unit tests pass locally with my changes (if applicable)
- [x] I have run pre-commit (`pre-commit` in the repo)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Linked to the relevant github issue or github discussion

## Additional testing

- Created a web bug token with a MS Teams webhook and checked that the webhook test message and token hit message were correct on MS Teams.
- Triggered an AWS API key token (using aws cli) and checked that the alert was sent to the specified email and generic webhook and that it displayed on the frontend.
- Triggered the exposed key alert with the following command

```
curl -X POST https://honeypdfs.net/$token --data-urlencode 'token_exposed=true' --data-urlencode 'exposed_time=1730710853' --data-urlencode 'public_location=https://google.com'
```

- Checked that the correct data was sent to the generic webhook

```
{
  "token_type": "aws_keys",
  "token": "... snip ...",
  "memo": "generic webhook test",
  "key_id": "... snip ...",
  "public_location": "https://google.com",
  "exposed_time": "2024-11-04 09:00:53 (UTC)",
  "manage_url": "https://honeypdfs.com/manage?token=... snip ...&auth=... snip ...",
  "public_domain": "honeypdfs.com"
}
```

- Checked that the data was displayed correctly on the frontend

CleanShot 2024-11-04 at 11 04 51@2x

### CODE REVIEW

File: canarytokens/webhook_formatting.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+    if isinstance(details, TokenAlertDetails):
+        return _format_alert_details_for_webhook(webhook_type, details)
+    else:
+        return _format_exposed_details_for_webhook(webhook_type, details)
+
+
+def _format_alert_details_for_webhook(
+    webhook_type: WebhookType, details: TokenAlertDetails
+):
+    if webhook_type == WebhookType.SLACK:
```
Comment (thinkst-tom) [inline canarytokens/webhook_formatting.py:80]: Could we use match and case here instead? Or are we not running python >= 3.10 here?

**Comment by thinkst-tom** [comment]
@gjcthinkst I think generalising tokens having alerts for both being exposed and hit is the right way to go. I know AWS has a way of finding out if a key is exposed and then alerts us (and then we alert users), but are there other tokens that we can have exposed alerts for?

**Review by thinkst-tom — APPROVED** [review]

File: canarytokens/webhook_formatting.py (hunk 1 above)
Comment (gjcthinkst) [reply canarytokens/webhook_formatting.py:80]: Unfortunately we still run Python 3.9 on our tokens servers so we can't use match just yet.

**Comment by gjcthinkst** [comment]
> @gjcthinkst I think generalising tokens having alerts for both being exposed and hit is the right way to go. I know AWS has a way of finding out if a key is exposed and then alerts us (and then we alert users), but are there other tokens that we can have exposed alerts for?

@thinkst-tom I'm not aware of other providers that also send us alerts like these currently, but I wouldn't be surprised if the other clouds add this in future as well.

**Pull request merged**